src/jit/morph.cpp

```diff
@@ -7058,180 +7086,82 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
-                hasMultiByteStackArgs =
-                    !VarTypeIsMultiByteAndCanEnreg(argx->TypeGet(), objClass, &typeSize, false, false);
-
-#if defined(UNIX_AMD64_ABI)
-                SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
+        calleeStackSize += argStackSize;
+        calleeFloatArgRegCount += argFloatRegCount;
+        calleeArgRegCount += argIntRegCount;
 
-                assert(objClass != nullptr);
-                eeGetSystemVAmd64PassStructInRegisterDescriptor(objClass, &structDesc);
-
-                if (structDesc.passedInRegisters)
-                {
-                    if (structDesc.eightByteCount == 2)
-                    {
-                        hasTwoSlotSizedStruct = true;
-                    }
-
-                    for (unsigned int i = 0; i < structDesc.eightByteCount; i++)
-                    {
-                        if (structDesc.IsIntegralSlot(i))
-                        {
-                            ++calleeArgRegCount;
-                        }
-                        else if (structDesc.IsSseSlot(i))
-                        {
-                            ++calleeFloatArgRegCount;
-                        }
-                        else
-                        {
-                            assert(false && "Invalid eightbyte classification type.");
-                            break;
-                        }
-                    }
-                }
-                else
-                {
-                    calleeStackSize += roundUp(typeSize, TARGET_POINTER_SIZE);
-                    hasMultiByteStackArgs = true;
-                }
-
-#elif defined(_TARGET_ARM64_) // ARM64
-                var_types hfaType  = GetHfaType(argx);
-                bool      isHfaArg = varTypeIsValidHfaType(hfaType);
-                size_t    size     = 1;
-
-                if (isHfaArg)
-                {
-                    hasHfaArg = true;
-
-                    calleeFloatArgRegCount += GetHfaCount(argx);
-                }
-                else
-                {
-                    // Structs are either passed in 1 or 2 (64-bit) slots
-                    size_t roundupSize = roundUp(typeSize, TARGET_POINTER_SIZE);
-                    size               = roundupSize / TARGET_POINTER_SIZE;
-
-                    if (size > 2)
-                    {
-                        size = 1;
-                    }
-
-                    else if (size == 2)
-                    {
-                        hasTwoSlotSizedStruct = true;
-                    }
-
-                    calleeArgRegCount += size;
-                }
-
-#elif defined(WINDOWS_AMD64_ABI)
-
-                ++calleeArgRegCount;
-
-#endif // UNIX_AMD64_ABI
+        // This exists to account for special case situations where we will not
+        // fast tail call.
+        if (arg->isStruct)
+        {
+            hasNonEnregisterableStructs = argStackSize > 0 ? true : hasNonEnregisterableStructs;
```

Comment:
IMO, it's better to write:
```
x = cond ? true : x;
```
as:
```
if (cond)
{
  x = true;
}
```
so you don't have to do a mental comparison of the two occurrences of `x` to determine that the "else" case is a no-op.